include/deal.II/lac/diagonal_matrix.h

```diff
@@ -196,6 +196,18 @@ class DiagonalMatrix : public Subscriptor
   void
   Tvmult_add(VectorType &dst, const VectorType &src) const;
 
+  /**
+   * Apply the preconditioner only to a subrange of elements of the given
+   * vector. To support this operation, the given `VectorType` template
+   * argument needs to support a method `begin()` to return the pointer to the
+   * start of the stored elements.
+   */
+  void
+  apply_to_subrange(const unsigned int index_of_first_unknown,
+                    const unsigned int length,
+                    const value_type * src,
+                    value_type *       dst) const;
```

Comment:
You are right, these are confusing. The problem that I am facing here is that the pointers need to be accessed in the interval `[0, length)`, because those might be local arrays which cannot be index as a whole vector. On the other hand, the preconditioner really needs to know the actual index it is working on (index in the MPI-local numbering scheme). If it weren't for the pointers, the most natural thing would be to specify `begin_range` and `end_range`, was we require in the matrix interface. Then we need to document clearly that the pointers pass in to this function are already "shifted" by `begin_range`, so the access to those vectors needs to go to the index range `[0, end_range - begin_range)`. Do you agree?

Reply:
Yes, I agree!